EventFilter/CastorRawToDigi/src/ZdcUnpacker.cc

```diff
@@ -0,0 +1,330 @@
+  myEMap[CastorElectronicsId(0,2,1,3,18,8,0)]=0x54000014;//NZDC EM4
+  myEMap[CastorElectronicsId(1,2,1,3,18,8,0)]=0x54000015;//NZDC EM5
+  myEMap[CastorElectronicsId(2,2,1,3,18,8,0)]=0x54000021;//NZDC HAD1
+  myEMap[CastorElectronicsId(0,3,1,3,18,8,0)]=0x54000022;//NZDC HAD2
+  myEMap[CastorElectronicsId(1,3,1,3,18,8,0)]=0x54000023;//NZDC HAD3
+  myEMap[CastorElectronicsId(2,3,1,3,18,8,0)]=0x54000024;//NZDC HAD4
+
+  //slot is 17
+
+  for (int spigot=0; spigot<HcalDCCHeader::SPIGOT_COUNT; spigot++) {
+    if (!dccHeader->getSpigotPresent(spigot)) continue;
+
+    int retval=dccHeader->getSpigotData(spigot,htr,raw.size());
+    if (retval!=0) {
+      if (retval==-1) {
+        if (!silent) edm::LogWarning("Invalid Data") << "Invalid HTR data (data beyond payload size) observed on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+        report.countSpigotFormatError();
+      }
+      continue;
+    }
+    // check
+    if (dccHeader->getSpigotCRCError(spigot)) {
+      if (!silent)
+        edm::LogWarning("Invalid Data") << "CRC Error on HTR data observed on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+      report.countSpigotFormatError();
+      continue;
+    }
+    if (!htr.check()) {
+      if (!silent)
+        edm::LogWarning("Invalid Data") << "Invalid HTR data observed on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+      report.countSpigotFormatError();
+      continue;
+    }
+    if (htr.isHistogramEvent()) {
+      if (!silent)
+        edm::LogWarning("Invalid Data") << "Histogram data passed to non-histogram unpacker on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId();
+      continue;
+
+    }
+    if ((htr.getFirmwareFlavor()&0xE0)==0x80) { // some kind of TTP data
+      if (colls.ttp!=0) {
+        HcalTTPUnpacker ttpUnpack;
+        colls.ttp->push_back(HcalTTPDigi());
+        ttpUnpack.unpack(htr,colls.ttp->back());
+      } else {
+        LogDebug("ZdcUnpackerHcalTechTrigProcessor") << "Skipping data on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId() << " which is from the TechTrigProcessor (use separate unpacker!)";
+      }
+      continue;
+    }
+    if (htr.getFirmwareFlavor()>=0x80) {
+      if (!silent) edm::LogWarning("ZdcUnpacker") << "Skipping data on spigot " << spigot << " of DCC with source id " << dccHeader->getSourceId() << " which is of unknown flavor " << htr.getFirmwareFlavor();
+      continue;
+    }
+    // calculate "real" number of presamples
+    //int nps=htr.getNPS()-startSample_;
+
+    // get pointers
+    htr.dataPointers(&daq_first,&daq_last,&tp_first,&tp_last);
+    unsigned int smid=htr.getSubmodule();
+    int htr_tb=smid&0x1;
+    int htr_slot=(smid>>1)&0x1F;
+    int htr_cr=(smid>>6)&0x1F;
+
+    //////////////////////////////////////////////
+    qie_begin=(HcalQIESample*)daq_first;
```

Comment:
this is still an issue based on the static code analyzer.
You have declared qie_begin and daq_first as const, but make a cast to a non-const here.
if this is not needed, better use const (both here and on L162/line below)